### ISSUE TITLE
Manual is hard to understand on how to resolve false positives

### ISSUE THREAD

**Issue by xenoterracide** [typetools/checker-framework#3991]
@mernst continuing from https://github.com/kelloggm/checkerframework-gradle-plugin/issues/109

> Can you suggest how the CF documentation can be improved, so that others don't face the same problem you did? Thanks.

these are the problems I had

https://stackoverflow.com/q/65245465/206466
https://stackoverflow.com/q/65245508/206466 it was incredibly non obvious how to use stubs given the documentation in the manual 
>   javac -processor org.checkerframework.checker.interning.InterningChecker -Astubs=String.astub:stubs MyFile.java MyOtherFile.java ...

what? maybe I did it wrong... see my SO answer though, doing that was not clear from this
https://stackoverflow.com/q/65224314/206466 I still don't understand why `Objects.requireNonNull` causes this error in the first place, the bad answer (because the solution didn't work at all), says it's because the NullChecker tries to prevent *all* null pointer exceptions. if that's the case if you search your documentation for `NullPointerException` it doesn't say that. If it's because the jdk stub (which I've been told is auto included in the gradle plugin), has `Objects.requireNonNull` as taking `@NonNull` then that is wrong. I didn't find a way to fix this outside of writing my own `Objects` stub. which seemed like the wrong answer for a jdk class.

this doesn't work, and as a side note, if it should work, there should probably be a task added in the gradle plugin that's documented

```sh
  java -cp $CHECKERFRAMEWORK/checker/dist/checker.jar org.checkerframework.framework.stub.JavaStubifier project-stubs
```

```sh
> java -cp /Users/calebcushing/.gradle/caches/modules-2/files-2.1/org.checkerframework/checker/3.8.0/b228a4b8777c157a68cc99b827048aa4a14b123a/checker-3.8.0.jar org.checkerframework.framework.stub.JavaStubifier project-stubs
> ls -l project-stubs                                          # ppm/scaf -> 75-migrations ! ?
ls: project-stubs: No such file or directory

####
from rereading it I didn't completely follow the instructions still am not
###
> mkdir project-stubs                                     # ppm/scaf -> 75-migrations ! ? RC=1
 > java -cp /Users/calebcushing/.gradle/caches/modules-2/files-2.1/org.checkerframework/checker/3.8.0/b228a4b8777c157a68cc99b827048aa4a14b123a/checker-3.8.0.jar org.checkerframework.framework.stub.JavaStubifier project-stubs
> ls project-stubs                                             # ppm/scaf -> 75-migrations ! ?                                                            
```
I got some error last night about a class missing, or having the wrong missing/method or something. Can't figure out what I did now.

honestly, instead of copying the source dir, you should be able to specify it, and checker should create the path to the destination, or throw a reasonable error if stuff doesn't exist.

**Comment by mernst** [comment]
Thanks for the feedback.  We appreciate it, and I'm sorry you had trouble.

> it was incredibly non obvious how to use stubs given the documentation in the manual
> > `javac -processor org.checkerframework.checker.interning.InterningChecker -Astubs=String.astub:stubs MyFile.java MyOtherFile.java ...`

Can you please clarify *what* was confusing?  I can't tell what to change from your remark.  Can you suggest a specific change to wording?

Your SO question shows you forgot to provide the path to the file.  I already [added text](https://github.com/typetools/checker-framework/commit/66ee97ff6174a69e7e65c87e37ac29400c5c173c) to try to prevent that confusion, so thanks for prompting that.  The very end of your text suggests issuing a warning when the user mis-specifies a file, which is a good idea and is now tracked at https://github.com/typetools/checker-framework/issues/3993.  Thanks for the suggestion.

> the NullChecker tries to prevent all null pointer exceptions. if that's the case if you search your documentation for `NullPointerException` it doesn't say that.

The [first sentence](https://checkerframework.org/manual/#nullness-checker) of the Nullness Checker documentation says, "If the Nullness Checker issues no warnings for a given program, then running that program will never throw a null pointer exception."
I [added the string `NullPointerException` to the text](https://github.com/typetools/checker-framework/commit/6b509e13eae5b0d2e2afb84ea961143f7507740f), for people who want to search for `NullPointerException`.  Thanks for the suggestion about adding that string.

> If it's because the jdk stub (which I've been told is auto included in the gradle plugin), has `Objects.requireNonNull` as taking `@NonNull` then that is wrong.

Can you please clarify what is wrong about that?  I guess you disagree with the section [Annotations indicate non-exceptional behavior](https://checkerframework.org/manual/#annotate-normal-behavior).  If you could open an issue just about what parts of that manual section you find wrong/confusing, that will help us to improve it or to change the Checker Framework behavior.  Thanks.

> I got some error last night about a class missing, or having the wrong missing/method or something. Can't figure out what I did now.

I'm sorry to hear that.  If you are able to produce more information, we will try to help.

> honestly, instead of copying the source dir, you should be able to specify it, and checker should create the path to the destination, or throw a reasonable error if stuff doesn't exist.

Regarding the last phrase, that is a good idea.  I have opened https://github.com/typetools/checker-framework/issues/3993 to track it.  Thanks.

Regarding the first parts of the sentence, I don't think it is necessary to copy a source directory.  Could you clarify your concern and your proposed fix, so that we can implement it?

I think this addresses all the different concerns you raised.  Thanks again.

**Comment by xenoterracide** [comment]
> Can you please clarify what was confusing?
 `-Astubs=String.astub:stubs` is String.astub:stubs what I need to add? how does that work? then I tried,  -`Astubs=String.astub:BooleanUtils` and a bunch of other incantations I can't remember. Eventually I gave up and just worked on figuring out how to make one by hand, which was also hard. The CF documentation didn't really give me a clue as to where to put this, how to structure the directories, etc. I eventually figured out where it was looking, and then adjusted my invocation accordingly.

> If the Nullness Checker issues no warnings for a given program, then running that program will never throw a null pointer exception.

You might be better off saying that 

> The Nullness Checker checks that the program does not throw any NullPointerException's via any means, including NullPointerException's from JDK calls,  you should replace any NullPointerException's with your own clearer exception.

> Can you please clarify what is wrong about that? I guess you disagree with the section Annotations indicate non-exceptional behavior.

I probably missed that section because I was jumping all over the manual. However, I do disagree with that point if the method is intentionally throwing an exception, even more so if like `Objects.requireNonNull` that is exactly what it is designed for. My opinion is that I recognize this is Nullable, and the program can't do anything but to terminate so I intentionally throw the exception. Wheras I do not want implicit NPE's, and that's what was my intent to look for.

> Regarding the first parts of the sentence, I don't think it is necessary to copy a source directory. Could you clarify your concern and your proposed fix, so that we can implement it?

the instructions say it's required, and it definitely didn't do anything without it, nor did I see instructions on how to simply point it to my source dir with an option. I can only assume that `src` refers to a well structured java folder (e.g. src/main/...)

```
 mkdir project-stubs
  cp -R project/src project-stubs
```

**Comment by xenoterracide** [comment]
@mernst any chance you could answer https://stackoverflow.com/q/65245465/206466 and maybe make the docs clearer on this, probably with another example? I'm thinking of reverting my position on using checker, but before I do I need to resolve this issue.

**Comment by xenoterracide** [comment]
Also, could you document in the stub's section how it's possible to override the jdk stubs? (this may be as obvious as simply providing one, I'm not certain).

**Comment by mernst** [comment]
> @mernst any chance you could answer https://stackoverflow.com/q/65245465/206466

Sure.  Could you please first provide a MWE?
For future reference, bug reports are usually better in the issue tracker; discussions on Stack Overflow get messy and are not very useful for others.

I'm also happy to improve the documentation.  It's better than most open source projects, but far from perfect.  We appreciate suggestions.

> Also I don't want to just open a ticket on it, but would it be possible to have a check that checks for proper nullness annotations existing on functions inside of your codebase? and fails on missing ones.

Sorry, I don't understand this.  Could you open a ticket, with a code example to help me understand?  Thank you.

**Comment by mernst** [comment]
> Also, could you document in the stub's section how it's possible to override the jdk stubs? (this may be as obvious as simply providing one, I'm not certain).

Good idea.

For future reference, please open a new ticket per distinct request, to make it easier for us to track.  Thanks!

**Comment by xenoterracide** [comment]
> discussions on Stack Overflow get messy and are not very useful for others.

discussions, yes, but is answering it a discussion, here's my full code, this reminds me to add it to the SO https://github.com/xenoterracide/brix/blob/master/src/main/java/com/xenoterracide/brix/Application.java it doesn't currently have NullnessCheck enabled, because things simply wouldn't compile with it.

> The manual explains that @NonNull is the correct annotation.

I mean, this seems like it's actually part of this request as it's documentation.

> Sorry, I don't understand this. Could you open a ticket, with a code example to help me understand? Thank you.

sure, tbh I wasn't sure if it was already possible via something.
